File: expression.java

// Time Complexity :exponential
// Space Complexity :O(n^2)
// Did this code successfully run on Leetcode :yes
// Any problem you faced while coding this :No


class Solution {
    List<String> result;

    public List<String> addOperators(String num, int target) {
        result = new ArrayList<>();
        helper(num, target, 0, 0, 0, "");
        return result;
    }

    private void helper(String num, int target, int idx, long calc, long tail, String path) {
        // base case//if we are at last index and found target, add in result and return
        if (idx == num.length()) {
            if (calc == target) {
                result.add(path);
            }
            return;
        }
        // logic//doing for loop recursion
        for (int i = idx; i < num.length(); i++) {
            // mking substring from idx to i to make digits
            long curr = Long.parseLong(num.substring(idx, i + 1));
            // if our digit created is not 0 but has preceding zero we dont make expresiion
            if (num.charAt(idx) == '0' && idx != i)
                continue;
            // at first index, we cannot make expression
            if (idx == 0) {
                helper(num, target, i + 1, curr, curr, path + curr);
            }
            // here we are taking cases of +,- and * for multiplication, because of BODMAS,
            // we need to maintain a tail
            // so that we can do calculation as precedence rule
            else {
                // +case
                helper(num, target, i + 1, calc + curr, curr, path + "+" + curr);
                // -case
                helper(num, target, i + 1, calc - curr, -curr, path + "-" + curr);
                // *case
                helper(num, target, i + 1, calc - tail + curr * tail, curr * tail, path + "*" + curr);
            }
        }
    }
}

File: Expression.py
"""
Time - O(N ^ ((Target / Minimum) + 1))
Space - O(N ^ ((Target / Minimum) + 1))
Leetcode - Yes
Problems - No
"""
class Solution:
    def addOperators(self, num: str, target: int) -> List[str]:
        if not num:
            return []
			
        num_len = len(num)
        res = []
        
        def dfs(idx, path):
            if idx == num_len - 1:
                # num_len - 1 to prevent situations like (1+2+) or (1*3*) etc..
                path += num[idx] 
                if eval(path) == target:
                    res.append(path)
                return
            
            dfs(idx+1, path + num[idx] + "+") 
            dfs(idx+1, path + num[idx] + "-")
            dfs(idx+1, path + num[idx] + "*")
            if (path and path[-1] not in ['+', '-', '*'] and num[idx] == '0') or num[idx] != '0':
                
                # Prevent cases such as (2+05) which cannot be evaluated using the eval function
                # Cases such as 12+104 should be acceptable, but not 121+04
                
                dfs(idx+1, path + num[idx])
            
        dfs(0, "")
        return res
    